extensions/amp-content-viewer/0.1/amp-content-viewer.js

```diff
@@ -0,0 +1,696 @@
+          e.data.deltaX, e.data.deltaY, e.data.dir);
+      if (e.data.last) {
+        this.onZoomRelease_();
+      }
+    });
+
+    // Override all taps
+    this.gestures_.onGesture(TapRecognizer, e => {
+      const event = createCustomEvent(this.win, 'click', null, {bubbles: true});
+      e.data.target.dispatchEvent(event);
+    });
+  }
+
+  /**
+   * Registers a Swipe gesture to handle panning when the content is zoomed.
+   * @private
+   */
+  registerPanningGesture_() {
+    // Movable.
+    this.unlistenOnSwipePan_ = this.gestures_
+        .onGesture(SwipeXYRecognizer, e => {
+          event.preventDefault();
+          this.onMove_(e.data.deltaX, e.data.deltaY, false);
+          if (e.data.last) {
+            this.onMoveRelease_(e.data.velocityX, e.data.velocityY);
+          }
+        });
+  }
+
+  /**
+   * Deregisters the Swipe gesture for panning when the content is zoomed out.
+   * @private
+   */
+  unregisterPanningGesture_() {
+    if (this.unlistenOnSwipePan_) {
+      this.unlistenOnSwipePan_();
+      this.unlistenOnSwipePan_ = null;
+      this.gestures_.removeGesture(SwipeXYRecognizer);
+    }
+  }
+
+  /**
+   * Returns value bound to min and max values +/- extent.
+   * @param {number} v
+   * @param {number} min
+   * @param {number} max
+   * @param {number} extent
+   * @return {number}
+   * @private
+   */
+  boundValue_(v, min, max, extent) {
+    return Math.max(min - extent, Math.min(max + extent, v));
+  }
+
+  /**
+   * Returns the scale within the allowed range with possible extent.
+   * @param {number} s
+   * @param {boolean} allowExtent
+   * @return {number}
+   * @private
+   */
+  boundScale_(s, allowExtent) {
+    return this.boundValue_(s, this.minScale_, this.maxScale_,
+        allowExtent ? 0.25 : 0);
+  }
+
+  /**
+   * Returns the X position within the allowed range with possible extent.
+   * @param {number} x
+   * @param {boolean} allowExtent
+   * @return {number}
+   * @private
+   */
+  boundX_(x, allowExtent) {
+    return this.boundValue_(x, this.minX_, this.maxX_,
+        allowExtent && this.scale_ > 1 ? this.elementBox_.width * 0.25 : 0);
+  }
+
+  /**
+   * Returns the Y position within the allowed range with possible extent.
+   * @param {number} y
+   * @param {boolean} allowExtent
+   * @return {number}
+   * @private
+   */
+  boundY_(y, allowExtent) {
+    return this.boundValue_(y, this.minY_, this.maxY_,
+        allowExtent ? this.elementBox_.height * 0.25 : 0);
+  }
+
+  /**
+   * Updates X/Y bounds based on the provided scale value. The min/max bounds
+   * are calculated to allow full pan of the content regardless of the scale
+   * value.
+   * @param {number} scale
+   * @private
+   */
+  updatePanZoomBounds_(scale) {
+    let maxY = 0;
+    let minY = 0;
+    const dh = this.elementBox_.height - this.contentBox_.height * scale;
+    if (dh >= 0) {
+      minY = maxY = 0;
+    } else {
+      minY = dh / 2;
+      maxY = -minY;
+    }
```

Comment:
nit: Ternaries can make this more declarative and easier to grok

```js
const dh = this.elementBox_.height - this.contentBox_.height * scale;
const minY = (dh >= 0) ? 0 : dh / 2;
const maxY = -minY;
```

Reply:
Done